### ISSUE TITLE
getmusic training dosnt export checkpoint file

### ISSUE THREAD

**Issue by kotaxyz** [microsoft/muzic#126]
hi there,i was able to run training but i cant find the checkpoint in the folder here is the log

(getmusic) C:\Users\Genesis\anaconda3\muzic\getmusic>python train.py
Global seed set to 42
args.num_node  1
Namespace(amp=False, auto_resume=False, config_file='configs/train.yaml', cudnn_deterministic=False, cwd='C:\\Users\\Genesis\\anaconda3\\muzic\\getmusic', debug=False, dist_url='tcp://127.0.0.1:29500', distributed=False, do_sample=False, global_rank=0, gpu=None, load_path=None, local_rank=0, log_frequency=20, name='train', ngpus_per_node=1, no_load_optimizer_and_scheduler=True, no_load_others=True, node_rank=0, num_node=1, opts=[], output='OUTPUT', resume_name=None, save_dir='OUTPUT\\train\\2023-05-26T17-39-07', seed=42, sync_bn=False, tensorboard=False, timestamp=False, world_size=1)
<class 'getmusic.modeling.models.dfm.DFM'>
<class 'getmusic.modeling.roformer.diffusion_roformer.DiffusionRFM'>
<class 'getmusic.modeling.roformer.roformer_utils.DiffusionRoformerModel'>
<class 'getmusic.data.bigdata.BigDataset'>
<class 'getmusic.data.bigdata.BigDataset'>
Load dictionary: 924 tokens.
<class 'getmusic.engine.clip_grad_norm.ClipGradNorm'>
Get lr 3e-06 from base lr 3e-06 with none
<class 'torch.optim.adamw.AdamW'>
<class 'getmusic.engine.lr_scheduler.LinearDecayLRWithWarmup'>
{'overall': {'trainable': '85.87M', 'non_trainable': '96.0K', 'total': '85.97M'}, 'rfm': {'trainable': '85.87M', 'non_trainable': '96.0K', 'total': '85.97M'}}
self.device  0
train: global rank 0: prepare solver done!
train: global rank 0: start training...
time2 is 1685111989.6986997
train: train: Epoch 0/50 iter 0/9 || | loss: 12.5147 || lr_lr: 3.097e-06 || data_time: 3.1s | fbward_time: 2.2s | iter_time: 5.3s | iter_avg_time: 5.3s | epoch_time: 05s | spend_time: 05s | left_time: 39m:39s
time2 is 1685111996.385699
time2 is 1685112001.431699
train: train: Epoch 2/50 iter 2/9 || | loss: 2.3620 || lr_lr: 5.037e-06 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.8s | epoch_time: 01s | spend_time: 16s | left_time: 05m:37s
time2 is 1685112006.484697
time2 is 1685112011.542714
train: train: Epoch 4/50 iter 4/9 || | loss: 4.6332 || lr_lr: 6.977e-06 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.7s | epoch_time: 02s | spend_time: 27s | left_time: 04m:36s
time2 is 1685112016.645698
time2 is 1685112021.7306979
train: train: Epoch 6/50 iter 6/9 || | loss: 4.3176 || lr_lr: 8.917e-06 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 03s | spend_time: 39s | left_time: 04m:09s
time2 is 1685112026.7956984
time2 is 1685112031.8566992
train: train: Epoch 8/50 iter 8/9 || | loss: 2.1051 || lr_lr: 1.0857e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 05s | spend_time: 50s | left_time: 03m:49s
time2 is 1685112036.9217083
time2 is 1685112041.989699
time2 is 1685112047.070699
train: train: Epoch 11/50 iter 1/9 || | loss: 0.4293 || lr_lr: 1.2797e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 01s | spend_time: 01m:01s | left_time: 03m:32s
time2 is 1685112052.1566994
time2 is 1685112057.2157094
train: train: Epoch 13/50 iter 3/9 || | loss: 0.4865 || lr_lr: 1.4737e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 02s | spend_time: 01m:12s | left_time: 03m:18s
time2 is 1685112062.3216991
time2 is 1685112067.3836997
train: train: Epoch 15/50 iter 5/9 || | loss: 0.3555 || lr_lr: 1.6677e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 03s | spend_time: 01m:24s | left_time: 03m:04s
time2 is 1685112072.414699
time2 is 1685112077.4276989
train: train: Epoch 17/50 iter 7/9 || | loss: 0.1403 || lr_lr: 1.8617e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 04s | spend_time: 01m:35s | left_time: 02m:51s
time2 is 1685112082.4526987
time2 is 1685112087.4656994
time2 is 1685112092.4846976
train: train: Epoch 20/50 iter 0/9 || | loss: 3.1592 || lr_lr: 2.0557e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 00s | spend_time: 01m:46s | left_time: 02m:38s
time2 is 1685112097.4966981
time2 is 1685112102.5276992
train: train: Epoch 22/50 iter 2/9 || | loss: 0.3356 || lr_lr: 2.2497e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 01s | spend_time: 01m:57s | left_time: 02m:25s
time2 is 1685112107.582699
time2 is 1685112112.6366992
train: train: Epoch 24/50 iter 4/9 || | loss: 0.7149 || lr_lr: 2.4437e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 02s | spend_time: 02m:08s | left_time: 02m:13s
time2 is 1685112117.663698
time2 is 1685112122.6986985
train: train: Epoch 26/50 iter 6/9 || | loss: 0.8707 || lr_lr: 2.6377e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 03s | spend_time: 02m:20s | left_time: 02m:01s
time2 is 1685112127.7446988
time2 is 1685112132.797698
train: train: Epoch 28/50 iter 8/9 || | loss: 0.0863 || lr_lr: 2.8317e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 05s | spend_time: 02m:31s | left_time: 01m:49s
time2 is 1685112137.8637002
time2 is 1685112142.9106987
time2 is 1685112147.9427137
train: train: Epoch 31/50 iter 1/9 || | loss: 1.0531 || lr_lr: 3.0257e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 01s | spend_time: 02m:42s | left_time: 01m:37s
time2 is 1685112152.9616985
time2 is 1685112157.9786985
train: train: Epoch 33/50 iter 3/9 || | loss: 1.3936 || lr_lr: 3.2197e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 02s | spend_time: 02m:53s | left_time: 01m:25s
time2 is 1685112163.052698
time2 is 1685112168.1066992
train: train: Epoch 35/50 iter 5/9 || | loss: 1.5818 || lr_lr: 3.4137e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 03s | spend_time: 03m:04s | left_time: 01m:14s
time2 is 1685112173.1206985
time2 is 1685112178.1657178
train: train: Epoch 37/50 iter 7/9 || | loss: 1.0043 || lr_lr: 3.6077e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 04s | spend_time: 03m:16s | left_time: 01m:02s
time2 is 1685112183.224697
time2 is 1685112188.282699
time2 is 1685112193.3316987
train: train: Epoch 40/50 iter 0/9 || | loss: 2.4089 || lr_lr: 3.8017e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 00s | spend_time: 03m:27s | left_time: 51s
time2 is 1685112198.3916988
time2 is 1685112203.4246979
train: train: Epoch 42/50 iter 2/9 || | loss: 0.9107 || lr_lr: 3.9957e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 01s | spend_time: 03m:38s | left_time: 39s
time2 is 1685112208.4466984
time2 is 1685112213.481707
train: train: Epoch 44/50 iter 4/9 || | loss: 0.1852 || lr_lr: 4.1897e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 02s | spend_time: 03m:49s | left_time: 28s
time2 is 1685112218.518698
time2 is 1685112223.5316982
train: train: Epoch 46/50 iter 6/9 || | loss: 0.5430 || lr_lr: 4.3837e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 03s | spend_time: 04m:00s | left_time: 16s
time2 is 1685112228.5496995
time2 is 1685112233.558698
train: train: Epoch 48/50 iter 8/9 || | loss: 0.0844 || lr_lr: 4.5777e-05 || data_time: 0.0s | fbward_time: 0.6s | iter_time: 0.6s | iter_avg_time: 0.6s | epoch_time: 05s | spend_time: 04m:11s | left_time: 05s
time2 is 1685112238.613698

here is my train.yaml settings 
i added  (save_checkpoints: True  - checkpoint_dir:  ) to the solver section in the train.yaml

 

model:
  target: getmusic.modeling.models.dfm.DFM
  params:
    diffusion_config:
      target: getmusic.modeling.roformer.diffusion_roformer.DiffusionRFM
      params:
        diffusion_step: 50
        alpha_init_type: 'alpha1'
        auxiliary_loss_weight: 0.001
        adaptive_auxiliary_loss: True
        roformer_config:
          target: getmusic.modeling.roformer.roformer_utils.DiffusionRoformerModel
          params:
            vocab_size: 5142
            cond_weight: 0.5

solver:
  base_lr: 3.0e-6
  adjust_lr: none
  max_epochs: 10
  save_epochs: 5
  validation_epochs: 1
  sample_iterations: epoch
  validate_iterations: 1000
  vocab_path: example_data/processed_train/pitch_dict.txt
  print_specific_things: True
  save_checkpoints: True
  checkpoint_dir: C:/Users/Genesis/anaconda3/muzic/getmusic/mymodels
  # config for ema
  ema:
   decay: 0.90
   update_interval: 100
   device: cuda

  clip_grad_norm:
    target: getmusic.engine.clip_grad_norm.ClipGradNorm
    params:
      start_iteration: 0
      end_iteration: 5000
      max_norm: 0.5
  optimizers_and_schedulers:
  - name: none
    optimizer:
      target: torch.optim.AdamW
      step_iteration: 1
      params:
        betas: !!python/tuple [0.9, 0.999]
        weight_decay: 1.0e-2
            
    scheduler:
      step_iteration: 1
      target: getmusic.engine.lr_scheduler.LinearDecayLRWithWarmup
      params:
        min_lr: 1.0e-6
        warmup_lr: 1.0e-4 # the lr to be touched after warmup
        warmup: 1000
        T_max: 300000

dataloader:
  batch_size: 1 # batch size per GPU
  num_workers: 4
  train_datasets:
    - target: getmusic.data.bigdata.BigDataset
      params:
        prefix: train
        path: C:\Users\Genesis\anaconda3\muzic\getmusic\example_data\processed_train\ # can be ignored if you use our checkpoints for just inference
        vocab_size: 5142
  validation_datasets:
    - target: getmusic.data.bigdata.BigDataset
      params:
        prefix: valid
        path: C:\Users\Genesis\anaconda3\muzic\getmusic\example_data\processed_train\
        vocab_size: 5142

![v](https://github.com/microsoft/muzic/assets/105466290/175de223-8182-41aa-a122-46805a766ed5)
![vv](https://github.com/microsoft/muzic/assets/105466290/76bcb5d1-be07-4b3e-b0a7-cd59cb64f41b)

**Comment by trestad** [comment]
Notice this part in the config file:

> solver:
  base_lr: 3.0e-6
  adjust_lr: none
  max_epochs: 50
  save_epochs: 10
  validation_epochs: 1
  sample_iterations: epoch
  validate_iterations: 1000
  vocab_path: getmusic/utils/dict.txt
  print_specific_things: True

In our default setting, the model validates and saves checkpoint every 1000 iterations. Since you do not have enough data, modify 'validate_iterations' to a smaller value.

**Comment by trestad** [comment]
We possess over 100,000 GETScores for training. We have yet to evaluate whether a set of 200 samples would be effective. It is recommended to fine-tune the hyperparameters, such as ema settings, learning rate, and batch size, through trial and error.

**Comment by 980202006** [comment]
@trestad I just started training the model. I have 70,000 available midi songs. Can you share your loss?

**Comment by trestad** [comment]
With 1k (2k, 4k, 8k, 16k, 32k) steps, the training loss is about 3 (2, 1, 0.9, 0.8, 0.6). When the training converges, the loss fluctuates around 0.4~0.5.

**Comment by 980202006** [comment]
@trestad Thank you for sharing, is the step you are talking about diffusion_step?Can you tell me your vocab_size? My vocab_size is 9165. Due to my limited video memory, about 19% of the chord vocabulary has been filtered by me.How long did you train the model to converge? My loss is still around 7~8.
